core/embed/trezorhal/image.c

```diff
@@ -50,35 +50,71 @@ static secbool compute_pubkey(uint8_t sig_m, uint8_t sig_n,
   return sectrue * (0 == ed25519_cosi_combine_publickeys(res, keys, sig_m));
 }
 
-secbool load_image_header(const uint8_t *const data, const uint32_t magic,
-                          const uint32_t maxsize, uint8_t key_m, uint8_t key_n,
-                          const uint8_t *const *keys, image_header *const hdr) {
+secbool parse_image_header(const uint8_t *const data, const uint32_t magic,
+                           const uint32_t maxsize, image_header *const hdr) {
   memcpy(&hdr->magic, data, 4);
   if (hdr->magic != magic) return secfalse;
 
   memcpy(&hdr->hdrlen, data + 4, 4);
   if (hdr->hdrlen != IMAGE_HEADER_SIZE) return secfalse;
 
   memcpy(&hdr->expiry, data + 8, 4);
+  memcpy(&hdr->codelen, data + 12, 4);
+  memcpy(&hdr->version, data + 16, 4);
+  memcpy(&hdr->fix_version, data + 20, 4);
+  memcpy(&hdr->monotonic, data + 24, 1);
+  memcpy(&hdr->hw_model, data + 25, 1);
+  memcpy(&hdr->hw_revision, data + 26, 1);
+  memcpy(hdr->hashes, data + 32, 512);
+  memcpy(&hdr->sigmask, data + IMAGE_HEADER_SIZE - IMAGE_SIG_SIZE, 1);
+  memcpy(hdr->sig, data + IMAGE_HEADER_SIZE - IMAGE_SIG_SIZE + 1,
+         IMAGE_SIG_SIZE - 1);
+
   // TODO: expiry mechanism needs to be ironed out before production or those
   // devices won't accept expiring bootloaders (due to boardloader write
   // protection).
-  if (hdr->expiry != 0) return secfalse;
+  if ((hdr->expiry & 0xFFFFFFFE) != 0) return secfalse;
+
+  // abusing expiry field to implement model check
```

Comment:
the comment is IMHO misleading because the point of the expiry hack is to prevent old bootloaders from running TR firmwares, not to model-check in new bootloaders.

i would modify the logic like so:
(a) if  `expiry == hw_model == hw_revision == 0` then allow for model T, disallow for others
(b) if `hw_model` and `hw_revision` match our data, then allow

(of course this would be much nicer if we were allowed an early return for true, but we probably don't want to do that here)

this additionally disallows `expiry == 1` and missing model info which is IMHO a good thing.
OTOH it allows `expiry == 0` and model info filled in, which seems ok

Reply:
`expiry == 0` is not acceptable for non-TT images, as older TT bootloaders that dont check `hw_model` will be able to run those images.

So :
`expiry == hw_model == hw_revision == 0 ` allow for TT, reject for others

`expiry != 0 && hw_model == HW_MODEL && hw_revision == HW_REVISION` allow for all models, though expiry will still be set to 0 on TT, to allow running on existing boardloaders/bootloaders

since we still want to fill hw_model and hw revision into new TT images, we also need to allow:
`expiry == 0 && hw_model == HW_MODEL && hw_revision == HW_REVISION` 

that is basically current logic, apart from the fact that current logic allows also combinations of hw_model == 0 and hw_revision == HW_REVISION and vice versa, which might not be desirable, but also shouldn't happen.

Also using just one bit of expiry to do this, in order to allow more functionality in the `expiry` field in the future.

Reply:
my point was that branching on the expiry bit seems to muddle the logic (to my reading at least)

the current code, and your explanation too, flows in _three_ distinct branches:
```
if expiry_bit is 0 {
   if model_t and hw_model is 0 and hw_revision is 0 {
       ok
   } else if hw_model is expected and hw_revision is expected {
       ok
   }
} else if hw_model is expected and hw_revision is expected {
   ok
}
```
the way I'm reading it, "expiry_bit is 0" is _not_ the deciding factor, and there are _two_ options: either the header is old-style TT, or it is new-style anything.
```
if model_t and expiry_bit is 0 and hw_model is 0 and hw_revision is 0 {
   ok  /* special case for TT preexisting images */
else if hw_model is expected and hw_revision is expected { 
   ok  /* general case for new images from today on */
}
```

Reply:
or are you saying that the bootloader on TR should explicitly reject images that have `hw_model == `R` && expiry == 0` ? if yes, why?

Reply:
because `expiry == 0` means TT image. So `hw_model == 'R' && expiry == 0` should never happen and something is broken

Reply:
the model_t condition is also valid for the `else if` branch

```
if expiry_bit is 0 {
   if model_t and hw_model is 0 and hw_revision is 0 {
       ok
   } else if model_t and hw_model is expected and hw_revision is expected {
       ok
   }
} else if hw_model is expected and hw_revision is expected {
   ok
}
```

Reply:
if the intention is to know that we don't produce `model == 'R' && expiry == 0` by mistake, a clearer way to write that would be to modify the `expiry & 0xfffe` on non-TT to explicitly check for 1

Reply:
modified the conditions: 
https://github.com/trezor/trezor-firmware/commit/43d87b70a075d1093e2b7c0c9a96c13acfad3d1a